# Import libraries 
library(atena)
library(SummarizedExperiment)
library(DESeq2)
library(tidyverse)
library(Rsamtools)
library(GenomicAlignments)
library(GenomicRanges)
library(data.table)

# Path to a BAM file
bam_file <- "14238-MF-0001_S1_L005_R1_001.fastq.gz.subread.BAM"

# Open BAM file
bam <- BamFile(bam_file)

# Read a small subset of reads
reads <- readGAlignments(bam, param=ScanBamParam(what=c("strand")))

# Count strands
table(strand(reads))
rm(bam, bam_file, reads)

# Results indicate this is unstranded library
# Proceed to TE alignment and quantification

# Download necessary data
#wget http://hgdownload.soe.ucsc.edu/goldenPath/hg38/database/rmsk.txt.gz
#gunzip rmsk.txt.gz

##########################################
# 1) Prepare BAM file list
##########################################
bam_files <- c(
  "14238-MF-0001_S1_L005_R1_001.fastq.gz.subread.BAM",
  "14238-MF-0002_S1_L005_R1_001.fastq.gz.subread.BAM",
  "14238-MF-0003_S1_L005_R1_001.fastq.gz.subread.BAM",
  "14238-MF-0004_S1_L005_R1_001.fastq.gz.subread.BAM",
  "14238-MF-0005_S1_L005_R1_001.fastq.gz.subread.BAM",
  "14238-MF-0006_S1_L005_R1_001.fastq.gz.subread.BAM"
)
bfl <- BamFileList(bam_files, yieldSize = 2000000)

##########################################
# 2) Import full UCSC RepeatMasker table
##########################################
# This assumes you have downloaded & unzipped rmsk.txt
rmsk <- fread("rmsk.txt", header = FALSE)

# Standard RepeatMasker columns (UCSC format)
colnames(rmsk) <- c(
  "bin", "swScore", "milliDiv", "milliDel", "milliIns",
  "genoName", "genoStart", "genoEnd", "genoLeft",
  "strand",
  "repName", "repClass", "repFamily",
  "repStart", "repEnd", "repLeft",
  "id"
)

# Build a GRanges with all metadata
rmsk_gr <- GRanges(
  seqnames = rmsk$genoName,
  ranges   = IRanges(start = rmsk$genoStart + 1, end = rmsk$genoEnd),
  strand   = rmsk$strand,
  swScore  = rmsk$swScore,
  milliDiv = rmsk$milliDiv,
  milliDel = rmsk$milliDel,
  milliIns = rmsk$milliIns,
  repName  = rmsk$repName,
  repClass = rmsk$repClass,
  repFamily= rmsk$repFamily,
  repStart = rmsk$repStart,
  repEnd   = rmsk$repEnd,
  repLeft  = rmsk$repLeft
)

##########################################
# 3) Parse RepeatMasker annotations
##########################################
parsed_rmsk <- rmskatenaparser(
  rmsk_gr,
  strict = FALSE,  # include a broad set of TEs
  insert = 1000    # merge fragments within 1000 bp
)

##########################################
# 4) Extract TE features by class
##########################################
te_features <- c(
  getLINEs(parsed_rmsk, relLength = 0.0),
  getSINEs(parsed_rmsk, relLength = 0.0),
  getLTRs(parsed_rmsk,  relLength = 0.0),
  getDNAtransposons(parsed_rmsk, relLength = 0.0)
)

##########################################
# 5) Build TEtranscripts parameter object
##########################################
te_param <- TEtranscriptsParam(
  bfl,
  teFeatures   = te_features,
  geneFeatures = NULL,   # omit gene counts
  ignoreStrand = TRUE    # unstranded libraries
)

##########################################
# 6) Sample metadata
##########################################
coldata <- data.frame(
  condition = factor(c(
    "P301L_V337M_DOX","P301L_V337M_DOX","P301L_V337M_DOX",
    "P301L_V337M",     "P301L_V337M",     "P301L_V337M"
  )),
  row.names = bam_files
)

##########################################
# 7) Quantify TE expression
##########################################
te_se <- qtex(
  te_param,
  phenodata = coldata
)

##########################################
# 8) Extract TE count matrix
##########################################
te_counts <- assay(te_se, "counts")

##########################################
# 9) DESeq2 differential analysis
##########################################
dds_te <- DESeqDataSetFromMatrix(
  countData = te_counts,
  colData   = coldata,
  design    = ~ condition
)

keep <- rowSums(counts(dds_te) >= 10) >= 3
dds_te <- dds_te[keep,]

dds_te <- DESeq(dds_te)

res_te <- results(
  dds_te,
  contrast = c("condition", "P301L_V337M_DOX", "P301L_V337M")
)

##########################################
# 10) Inspect top TE families
##########################################
res_sorted <- res_te[order(res_te$padj), ]
head(as.data.frame(res_sorted), 30)

# Save data
saveRDS(te_se, "2025.12.29_te_se.RDS")

# Convert results to data frame
res_df <- as.data.frame(res_te)
res_df$feature <- rownames(res_df)

# Derive a broad class from the feature name
res_df$TEclass <- with(res_df, ifelse(
  grepl("^Alu|^MIR", feature), "SINE",
  ifelse(grepl("^L1|^L2", feature), "LINE",
  ifelse(grepl("^ERV|^LTR", feature), "LTR",
  ifelse(grepl("^Charlie|^Tc", feature), "DNA",
         "Other")))))

# Make TEclass a factor
res_df$TEclass <- factor(res_df$TEclass,
                         levels = c("LINE", "SINE", "LTR", "DNA", "Other"))

# Assign colours for the broad classes
class_cols <- c(
  "LINE"   = "#1f77b4",
  "SINE"   = "#ff7f0e",
  "LTR"    = "#2ca02c",
  "DNA"    = "#d62728",
  "Other"  = "grey70"
)

#png("volcano_TE_by_family.png", width = 2500, height = 2400, res = 600)
# Draw the volcano plot, colouring by TE class
with(res_df, plot(
  log2FoldChange, -log10(padj),
  pch = 20,
  col = class_cols[as.character(TEclass)],
  xlab = "log2 Fold Change",
  ylab = "-log10 Adjusted P-value",
  main = "TE Differential Expression P301L,V337M vs P301L,V337M DOX"
))

# Add threshold lines
abline(v = c(-1, 1), col = "black", lty = 2)
abline(h = -log10(0.05), col = "black", lty = 2)

# Optional: add a legend
legend(
  "topright",
  legend = levels(res_df$TEclass),
  col = class_cols[levels(res_df$TEclass)],
  pch = 20,
  cex = 0.8,
  bty = "n"
)

# Highlight significant points (padj < 0.05 & |log2FC| > 1)
sig <- res_df$padj < 0.05 & abs(res_df$log2FoldChange) > 1
with(res_df[sig,], points(
  log2FoldChange, -log10(padj),
  pch = 20,
  col = class_cols[as.character(TEclass)]
))
#dev.off()

# Save output
write.csv(res_df, "2025.12.19_TE_DESeq2_results.csv")
